### PR TITLE
Add Tag Helper descriptors to the Tag Helper node

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Extensions
+{
+    internal class DefaultTagHelperOptimizationPass : RazorIRPassBase, IRazorIROptimizationPass
+    {
+        // Run later than default order so other passes have a chance to modify the
+        // tag helper nodes.
+        public override int Order => 1000;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs]: Hmm? DefaultFeatureOrder is 1k

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIRNode irDocument)
+        {
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs]: This class not needed?

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/ExtensionIRNode.cs [hunk 3]
```diff
@@ -36,6 +36,8 @@ public override RazorDiagnosticCollection Diagnostics
             }
         }
 
+        public override SourceSpan? Source { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/ExtensionIRNode.cs]: 👍

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/TagHelperIRNode.cs [hunk 4]
```diff
@@ -46,6 +47,8 @@ public override RazorDiagnosticCollection Diagnostics
 
         public TagMode TagMode { get; set; }
 
+        public ICollection<TagHelperDescriptor> TagHelpers { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/TagHelperIRNode.cs]: Thinking it's probably better to treat this guy like Children, Diagnostics, Annotations etc. and have a getter only.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/TagHelperIRNode.cs]: Sure

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs]: oh lol, that's not supposed to be in this commit

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs]: does it be?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs]: yeah I should be using the constant here. The default for user code is 0. The default for our code is 1000 where we want to run **after** user code. This is one of those cases

**Comment by NTaylorMullen** [comment]
Looks like you're failing in the travis/appveyor builds because of compile errors. :shipit: once everything be good.

File: test/Microsoft.AspNetCore.Razor.Test.Common/Langauge/Intermediate/RazorIRAssert.cs [hunk 5]
```diff
@@ -342,14 +342,16 @@ internal static void TagHelperFieldDeclaration(RazorIRNode node, params string[]
...
+        internal static void TagHelper(string tagName, TagMode tagMode, IEnumerable<TagHelperDescriptor> tagHelpers, RazorIRNode node, params Action<RazorIRNode>[] childValidators)
         {
             var tagHelperNode = Assert.IsType<TagHelperIRNode>(node);
 
             try
             {
                 Assert.Equal(tagName, tagHelperNode.TagName);
                 Assert.Equal(tagMode, tagHelperNode.TagMode);
+
+                Assert.Equal(tagHelpers, tagHelperNode.TagHelpers);
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Test.Common/Langauge/Intermediate/RazorIRAssert.cs]: @NTaylorMullen, should this use the case sensitive comparer?

Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Test.Common/Langauge/Intermediate/RazorIRAssert.cs]: You totally be right 👍

**Pull request merged**